tests/tcp_socket.rs

```diff
@@ -56,3 +56,39 @@ fn get_localaddr() {
 
     let _ = socket.listen(128).unwrap();
 }
+
+#[test]
+fn send_buffer_size_roundtrips() {
+    fn test(size: u32) {
+        println!("testing send buffer size: {}", size);
+        let socket = TcpSocket::new_v4().unwrap();
+        socket.set_send_buffer_size(size).unwrap();
+        // As per `man socket(7)`:
+        // > Sets or gets the maximum socket send buffer in bytes.  The
+        // > kernel doubles this value (to allow space for bookkeeping
+        // > overhead) when it is set using setsockopt(2), and this doubled
+        // > value is returned by getsockopt(2).
+        assert_eq!(size * 2, socket.get_send_buffer_size().unwrap());
+    }
+
+    test(4096);
+    test(65512);
+}
+
+#[test]
+fn recv_buffer_size_roundtrips() {
+    fn test(size: u32) {
+        println!("testing recv buffer size: {}", size);
+        let socket = TcpSocket::new_v4().unwrap();
+        socket.set_recv_buffer_size(size).unwrap();
+        // As per `man socket(7)`:
+        // > Sets or gets the maximum socket receive buffer in bytes.  The
+        // > kernel doubles this value (to allow space for bookkeeping
+        // > overhead) when it is set using setsockopt(2), and this doubled
+        // > value is returned by getsockopt(2).
```

Comment:
I'm not sure if this is also the case on Windows/BSD or not...guess we'll find out when CI fails.

Reply:
...cool, it looks like this [is, in fact, a Linux-specific behavior](https://dev.azure.com/tokio-rs/Tokio/_build/results?buildId=6430&view=logs&j=9ee66ee0-c6b6-50f2-32c8-2bebb82e1137&t=3b0adf12-f909-59d8-0849-13de19d8dc87&l=250)! Yay!

@Thomasdezeeuw, that brings up a question about what the right API to provide is. We could easily add a `#[cfg(target_os = "linux")]` in the tests and only expect a doubled value on Linux. However, since Mio's goal is to provide a cross-platform abstraction, would it make more sense to have the getters divide the value by 2 on Linux, so that the API behaves the same across all platforms? FWIW, I noticed that [`socket2` does _not_ have this behavior](https://github.com/rust-lang/socket2-rs/blob/b0f77842b3357dd571bb86b30b354bc55215f693/src/sys/unix.rs#L896-L901), but I'm not sure if that's relevant or not.

Reply:
You can change this to `socket.get_recv_buffer_size().unwrap() >= size` with a comment saying that Linux behaviour is different, as long as we have a check that the option is correctly applied I'm fine with it.